man/pjHinterp.1
.\" -*- NROFF -*-
.\" 
.TH "pjHinterp" "1" "" "Kevin Rosenberg and Ian Kay" "Engineering"
.SH "NAME"
pjHinterp \- interpolate helical data in projection space.
.SH "SYNOPSIS"
.B pjHinterp projfile interpfile [OPTIONS]
.SH "DESCRIPTION "
\fIpjHinterp\fP interpolates the helical data in projfile
and writes the results in interpfile.
.SH "OPTIONS"
.TP 16
.B \-\-field\-of\-view  
Field of view (ratio to diameter of phantom square) (default = 1)
.TP 16
.B \-\-trace          
Trace level to use, one of:
.RS 
.TP 
.B none 
No tracing (default)
.TP 16
.B console 
Trace text level
.RE
.TP 16
.B \-\-verbose        
Verbose mode
.TP 16
.B \-\-debug          
Debug mode
.TP 16
.B \-\-version        
Print version
.TP 16
.B \-\-help           
Print a terse help message

.SH "AUTHORS"
Ian Kay, Ph.D.
Kevin Rosenberg, M.D. <kevin@ctsim.org>
.SH "HISTORY"
CTSim was begun in 1983 using MS\-DOS and an EGA display adapter. In
1999 it was ported to GNU/Linux and later ported to Microsoft Windows.
.SH "SEE ALSO"
.BR ctsim (1)
.BR ctsimtext (1)
.BR if1 (1)
.BR if2 (1)
.BR ifinfo (1)
.BR linogram (1)
.BR phm2helix (1)
.BR phm2if (1)
.BR phm2pj (1)
.BR pj2if (1)
.BR pjinfo (1)
.BR pjrec (1)


NEWS
Version 4.2 Released Mar 2003

* Direct Fourier Reconstructions now supported

Version 4.0 Released Jan 2003

* Port to wxWindows 2.4

* Port to Mac OS X


Version 3.5 Released May 2002

* New reconstruction technique: Interpolate divergent beams to
  parallel beams for faster reconstructions.

* Multithreading for running length processes in background. Also
  takes advantage of multiple-CPU's in SMP systems for speed improvements
  with parallel processing.

* Region of interest can be specified for reconstruction zooming.

* Import PNG, PPM, and PGM images

* DICOM file import/export

* Clipboard for cutting and pasting images

* Can read and reconstruct SIEMENS AR.STAR DICOM raw data files

* Scattergram of T-Theta sampling. Useful for understanding divergent beam
  scanning.

* Startup-tips to help new users learn features.

* Ability to place detector at any arbitray position

* Histogram plotting of projection files


Version 3.0 New Features

* Greatly improved dialog boxes

* Online and print manuals with context-sensitive help!

* 3-d rotating views of image files

* Creation of arbitrary filter images

* Cubic interpolation for reconstructions

* All features of command line tools are now in graphical ctsim program!

* Complex-valued image files now supported, lots of image math
 functions added including Fourier transformations

* Visual and statistical image comparision functions

* Plotting of row and column data of single and comparison images

* Histogram Plotting of Image files.

* Conversion of projections to polar images

* Improved animation graphics

* New Microsoft Windows compatible self-installer with explorer extensions 
to CTSim files



Version 2.5 of CTSim has been released!

New Features of CTSim version 2.5:

- Now compiles with Microsoft Visual C++.

- First Microsoft Windows graphical user interface

- Bug fixes and modest documentation and program improvements over version 2.0



New Features of CTSim version 2.0:

- Entire code-base cleaned up, re-written, and converted from C to C++.

- Graphical user interface now included! Animation of projection and
  reconstruction processes. Displays projection and image files. Uses
  platform-independent wxWindows library so can run on UNIX and 
  Microsoft Windows platforms.

- Integrated G2 library for graphical display from command line tools
  (UNIX only)

- Added frequency-based filtering. Can provide significant speed-up
  in reconstructions compared to convolution-based filtering.

- Added highly-optimized backprojection method (idiff3)

- New command-line tools added for image statistics and comparisons with
  both text and graphical outputs.


tools/pjHinterp.cpp
/*****************************************************************************
 ** FILE IDENTIFICATION
 **
 **   Name:          pjHinterp.cpp
 **   Purpose:       Interpolate helical data in projection space
 **   Programmer:    Ian Kay and Kevin Rosenberg
 **   Date Started:  Aug 2001
 **
 **  This is part of the CTSim program
 **  Copyright (C) 1983-2009 Kevin Rosenberg
 **
 **  This program is free software; you can redistribute it and/or modify
 **  it under the terms of the GNU General Public License (version 2) as
 **  published by the Free Software Foundation.
 **
 **  This program is distributed in the hope that it will be useful,
 **  but WITHOUT ANY WARRANTY; without even the implied warranty of
 **  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 **  GNU General Public License for more details.
 **
 **  You should have received a copy of the GNU General Public License
 **  along with this program; if not, write to the Free Software
 **  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 ******************************************************************************/

#include "ct.h"
#include "timer.h"


enum { O_INTERPVIEW, O_VERBOSE, O_TRACE, O_HELP, O_DEBUG, O_VERSION};

static struct option my_options[] =
{
        {"interpview", 1, 0, O_INTERPVIEW},
        {"trace", 1, 0, O_TRACE},
        {"debug", 0, 0, O_DEBUG},
        {"verbose", 0, 0, O_VERBOSE},
        {"help", 0, 0, O_HELP},
        {"version", 0, 0, O_VERSION},
        {0, 0, 0, 0}
};

static const char* g_szIdStr = "$Id$";

void pjHinterp_usage ( const char *program )
{
  std::cout << "usage: " << fileBasename(program) << " projection-file interp-projection-file [OPTIONS]" << std::endl;
  std::cout << "Interpolation of helical data in raw data space" << std::endl;
  std::cout << "  projection-file Input projection file" << std::endl;
  std::cout << "  interp-file     Output interpolated projection file " << std::endl;
  std::cout << "  --trace         Set tracing to level" << std::endl;
  std::cout << "     none         No tracing (default)" << std::endl;
  std::cout << "     console      Text level tracing" << std::endl;
  std::cout << "  --verbose       Turn on verbose mode" << std::endl;
  std::cout << "  --debug         Turn on debug mode" << std::endl;
  std::cout << "  --version       Print version" << std::endl;
  std::cout << "  --help          Print this help message" << std::endl;
}


int
pjHinterp_main(int argc, char * const argv[])
{
  Projections projGlobal;
  char* pszProjFilename = NULL;
  char* pszInterpFilename = NULL;
  bool bOptVerbose = false;
  bool bOptDebug = 1;
  int optTrace = Trace::TRACE_NONE;
  char *endptr = NULL;
  char *endstr;
  int opt_interpview=-1;

  while (1) {
    int c = getopt_long(argc, argv, "", my_options, NULL);

    if (c == -1)
      break;

    switch (c) {
    case O_INTERPVIEW:
      opt_interpview = strtol(optarg, &endptr, 10);
      endstr = optarg + strlen(optarg);
      if (endptr != endstr) {
        std::cerr << "Error setting --interpview to %s" << optarg << std::endl;
        pjHinterp_usage(argv[0]);
        return(1);
      }
      break;
    case O_VERBOSE:
      bOptVerbose = true;
      break;
    case O_DEBUG:
      bOptDebug = true;
      break;
    case O_TRACE:
      if ((optTrace = Trace::convertTraceNameToID(optarg))
          == Trace::TRACE_INVALID) {
        pjHinterp_usage(argv[0]);
        return (1);
      }
      break;
    case O_VERSION:
#ifdef VERSION
      std::cout <<  "Version " <<  VERSION << std::endl <<
        g_szIdStr << std::endl;
#else
      std::cout << "Unknown version number" << std::endl;
#endif
      return (0);

    case O_HELP:
    case '?':
      pjHinterp_usage(argv[0]);
      return (0);
    default:
      pjHinterp_usage(argv[0]);
      return (1);
    } // end switch
  } // end while

  if (optind + 2 != argc) {
    pjHinterp_usage(argv[0]);
    return (1);
  }

  pszProjFilename = argv[optind];

  pszInterpFilename = argv[optind + 1];

  Projections projections;
  if ( projections.read(pszProjFilename) != true ){
    std::cerr << "Error reading input file " << pszProjFilename << std::endl;
    return (1);
  }
  if (bOptVerbose) {
          std::ostringstream os;
    projections.printScanInfo(os);
    std::cout << os.str();
  }

  int status = projections.Helical180LI(opt_interpview);
  if ( status != 0 )  return (1);
  status = projections.HalfScanFeather();
  if ( status != 0 )  return (1);
  projections.write( pszInterpFilename  );

  return (0);
}


#ifndef NO_MAIN
int
main (int argc, char* argv[])
{
  int retval = 1;

  try {
    retval = pjHinterp_main(argc, argv);
  } catch (exception e) {
      std::cerr << "Exception: " << e.what() << std::endl;
  } catch (...) {
      std::cerr << "Unknown exception" << std::endl;
  }

  return (retval);
}
#endif
